### PR TITLE
Estebanreyl/runtime sdk v2

### PR DESCRIPTION
**Pull request by estebanreyl**
Companion to https://github.com/Azure/azure-rest-api-specs/pull/6963 
Testing the second version of the Azure container registries SDK with built in blob paths (SDK is generated from https://github.com/AzureCR/azure-rest-api-specs/tree/estebanreyl/data-plane-08-2019)

This houses the newly generated paths as well as the test harness for the ACR dataplane runtime sdk.

Addresses: https://github.com/Azure/azure-rest-api-specs/issues/6773

### CODE REVIEW

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        private string AuthHeader { get; set; }
+        private LoginMode Mode { get; set; }
+        private string LoginUrl { get; set; }
+        private string Username { get; set; }
+        private string Password { get; set; }
+        private String Tenant { get; set; }
+        private CancellationToken RequestCancellationToken { get; set; }
+
+        // Structure : Scope : Token
+        private Dictionary<string, Token> AcrAccessTokens;
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs]: >AcrAccessTokens [](start = 42, length = 15)

camelCase all private fields if the SDK has a coding guidelines. Also can you elide the commonInit method and just initialize them here?

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        }
+
+        private void Print(HttpRequestMessage request)
+        {
+            Console.WriteLine("Method:{0} > {1}", request.Method, request.RequestUri);
+            Console.WriteLine("Headers: {0}", request.Headers.ToString());
+            Console.WriteLine("Content: ");
+            Console.WriteLine(request.Content?.ReadAsStringAsync().GetAwaiter().GetResult());
+            Console.WriteLine("");
+            //Console.WriteLine(request.Content?.Headers?.ToString());
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs]: > //Console.WriteLine(request.Content?.Headers?.ToString()); [](start = 11, length = 59)

Remove commented code or use Conditional Debug with a common method called PrintHeaders.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        }
+
+        #endregion
+
+        #region Helpers
+
+        /* Acquires a new ACR access token if necessary. It can also acquire a cached access token in order to avoid extra requests to
+         * the oauth2 endpoint improving efficiency. */
+        public string getAcrAccessToken(string scope)
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs]: >getAcrAccessToken [](start = 22, length = 17)

Check with the SDK team for coding guidelines here - PascalCase all methods.

**Comment by sajayantony** [comment]
Please check with the SDK team for coding guidelines - Maybe you can start with this- [https://github.com/dotnet/codeformatter](https://github.com/dotnet/codeformatter)

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+                {
+                    request.Headers.Authorization = new AuthenticationHeaderValue("Basic", AuthHeader);
+                }
+                await base.ProcessHttpRequestAsync(request, cancellationToken);
+            }
+        }
+
+        public struct Token
+        {
+            public string TokenStr { get; set; }
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs]: You can remove type suffixes. Rename to token.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        public static Token MakeToken(string token)
+        {
+            Token tok = new Token();
+            tok.TokenStr = token;
+            SecurityToken fields = JwtSecurityClient.ReadToken(token);
+            tok.Expiration = fields.ValidTo;
+            return tok;
+        }
+
+        public enum LoginMode
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs]: >LoginMode [](start = 20, length = 9)

Recommend moving to its own file outside the class

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+        {
+            Basic,
+            TokenAuth,
+            TokenAad
+        }
+
+        // Constant to refresh tokens slighly before they are to expire guarding against possible latency related crashes
+        public const double LATENCY_SAFETY = 2;
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs]: >LATENCY_SAFETY [](start = 28, length = 14)

suffix with units or use TimeSpans here.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+        /* Meant to parse out comma separated key value pairs delineated by the = sign. Accepts strings of the
+         format "key=value,key2=value2..." . Note this method is meant to provide limited functionality and
+         is not very robust. */
+        private Dictionary<string, string> parseHeader(string header)
+        {
+
+            Dictionary<string, string> parsed = new Dictionary<string, string>();
+
+            Regex re = new Regex(@"([\w\s]+)=""[^""]+"""); //Regex is required to allow multiple scopes like push,pull
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs]: >Regex re = new Regex(@"([\w\s]+)=""[^""]+""") [](start = 12, length = 45)

Define this as a static Regex outside the scope of the method and possibly compile at initialization time.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.ContainerRegistry
+{
+    /*Robust handling of Basic and OAUTH2 authentication flows for the Azure Container Registry Runtime .Net SDK.
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs]: >/* [](start = 4, length = 2)

NIT: use /// for comments. These are all minor C# specific guidelines so running the code formatter might help fix most of these.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs [hunk 9]
```diff
@@ -31,34 +32,45 @@ public static partial class AzureContainerRegistryClientExtensions
             /// <param name='cancellationToken'>
             /// The cancellation token.
             /// </param>
-            public static async Task GetDockerRegistryV2SupportAsync(this IAzureContainerRegistryClient operations, CancellationToken cancellationToken = default(CancellationToken))
+            public static async Task<object> CheckV2SupportAsync(this IAzureContainerRegistryClient operations, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs]: >CheckV2SupportAsync [](start = 45, length = 19)

I wonder if this API is even needed since we only support the V2 version.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs [hunk 10]
```diff
@@ -69,65 +81,355 @@ public static async Task<RepositoryTags> GetTagListAsync(this IAzureContainerReg
...
             /// A tag or a digest, pointing to a specific image
             /// </param>
+            /// <param name='payload'>
+            /// Manifest body, can take v1 or v2 values depending on accept header
+            /// </param>
             /// <param name='cancellationToken'>
             /// The cancellation token.
             /// </param>
-            public static async Task<Manifest> GetManifestAsync(this IAzureContainerRegistryClient operations, string name, string reference, CancellationToken cancellationToken = default(CancellationToken))
+            public static async Task<object> CreateManifestAsync(this IAzureContainerRegistryClient operations, string name, string reference, Manifest payload, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs]: >object [](start = 37, length = 6)

The earlier method returned a Manifest which might have been more useful in the chaining. Ideally having it return a a ManifestResponse which hopefully has the server side digest would be more useful.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs [hunk 11]
```diff
@@ -69,65 +81,355 @@ public static async Task<RepositoryTags> GetTagListAsync(this IAzureContainerReg
...
+            /// Name of the image (including the namespace)
+            /// </param>
+            /// <param name='digest'>
+            /// Digest of a BLOB
             /// </param>
             /// <param name='cancellationToken'>
             /// The cancellation token.
             /// </param>
-            public static async Task<Repositories> GetRepositoriesAsync(this IAzureContainerRegistryClient operations, string last = default(string), string n = default(string), CancellationToken cancellationToken = default(CancellationToken))
+            public static async Task<CheckBlobExistenceHeaders> CheckBlobExistenceAsync(this IAzureContainerRegistryClient operations, string name, string digest, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs]: >CheckBlobExistenceAsync [](start = 64, length = 23)

Recommendation for naming here would be GetBlobProperties.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs [hunk 12]
```diff
@@ -69,65 +81,355 @@ public static async Task<RepositoryTags> GetTagListAsync(this IAzureContainerReg
...
+            /// <param name='name'>
+            /// Name of the image (including the namespace)
+            /// </param>
+            /// <param name='digest'>
+            /// Digest of a BLOB
+            /// </param>
+            /// <param name='cancellationToken'>
+            /// The cancellation token.
+            /// </param>
+            public static async Task<Stream> DeleteBlobAsync(this IAzureContainerRegistryClient operations, string name, string digest, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs]: >DeleteBlobAsync [](start = 45, length = 15)

Can we remove this method. There is no client that would need this.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs [hunk 13]
```diff
@@ -69,65 +81,355 @@ public static async Task<RepositoryTags> GetTagListAsync(this IAzureContainerReg
...
+            /// <param name='_state'>
+            /// Acquired from NextLink
+            /// </param>
+            /// <param name='_nouploadcache'>
+            /// Acquired from NextLink
+            /// </param>
+            /// <param name='cancellationToken'>
+            /// The cancellation token.
+            /// </param>
+            public static async Task<GetBlobUploadStatusSpecifiedHeaders> GetBlobUploadStatusSpecifiedAsync(this IAzureContainerRegistryClient operations, string name, string uuid, string _state = default(string), bool? _nouploadcache = default(bool?), CancellationToken cancellationToken = default(CancellationToken))
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs]: >GetBlobUploadStatusSpecifiedAsync [](start = 74, length = 33)

Can we shorten this to GetBlobUploadStatus()

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs [hunk 14]
```diff
@@ -69,65 +81,355 @@ public static async Task<RepositoryTags> GetTagListAsync(this IAzureContainerReg
...
+            /// <param name='_state'>
+            /// Acquired from NextLink
+            /// </param>
+            /// <param name='_nouploadcache'>
+            /// Acquired from NextLink
+            /// </param>
+            /// <param name='cancellationToken'>
+            /// The cancellation token.
+            /// </param>
+            public static async Task<UploadBlobContentSpecifiedHeaders> UploadBlobContentSpecifiedAsync(this IAzureContainerRegistryClient operations, Stream value, string name, string uuid, string _state = default(string), bool? _nouploadcache = default(bool?), CancellationToken cancellationToken = default(CancellationToken))
```
Comment (sajayantony) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Generated/AzureContainerRegistryClientExtensions.cs]: >UploadBlobContentSpecifiedAsync [](start = 72, length = 31)

Same question - what is the value add of the long name ContentSpecified?

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs (hunk 5 above)
Comment (estebanreyl) [reply sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs]: I broke a lot of the token stuff apart into a separate class meant to handle tokens better.
